### PR TITLE
902 sql editor autocomplete

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes https://github.com/PopulateTools/issues/issues/902

## :v: What does this PR do?
Active autocomplete for SQL and columns from datasets

## :mag: How should this be manually tested?
Staging

### After this PR

autocomplete 2020-02-07 09_29_05

### CODE REVIEW

File: app/javascript/gobierto_data/lib/sql-theme.css [hunk 1]
```diff
@@ -0,0 +1,21 @@
+.sql {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/sql-theme.css:8]: Too simple name

File: app/javascript/gobierto_data/lib/sql-theme.css [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  color: #0000FF;
+}
+
+.sql::after,
+.table::after {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/sql-theme.css:13]: Too simple name

File: app/javascript/gobierto_data/lib/sql-theme.css [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  float: right;
+  color: #F0964D;
+}
+
+.table {
+  color: #8A2BE2;
+}
+
+.table::after {
+  content: 'tabla';
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/sql-theme.css]: Spanish?

File: app/javascript/gobierto_data/lib/sql-theme.css [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  color: #0000FF;
+}
+
+.sql::after,
+.table::after {
+  content: 'sql';
+  padding-left: 1.5rem;
+  font-style: italic;
+  float: right;
+  color: #F0964D;
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/sql-theme.css]: An straight color set here? With no variable?

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue [hunk 5]
```diff
@@ -1,61 +1,62 @@
 <template>
   <div>
     <div class="codemirror">
-      <codemirror
+      <textarea
         ref="myCm"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue]: "myCm", myCodemirror? use a proper name. What are you gonna type inside the textarea? A query? Then use names like `ref="queryEditor"`

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue [hunk 6]
```diff
@@ -1,61 +1,62 @@
 <template>
   <div>
     <div class="codemirror">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue]: Is this a required class for codemirror?

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue [hunk 7]
```diff
@@ -75,7 +79,7 @@ export default {
         mode: 'text/x-sql',
         hintOptions: {
           completeSingle: false,
-          tables: {}
+          hint: this.hint
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue]: What is `this.hint`? Looking the code is null. Do not initialize `data` variables with another `data` variable value. Perhaps a `prop`, but we should use simple types

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue [hunk 8]
```diff
@@ -108,32 +112,44 @@ export default {
...
+      editor.showHint()
+      this.$root.$emit('activeSave', false);
+      if (this.saveQueryState === true) {
+        this.$root.$emit('updateActiveSave', true, false);
       }
     })
-    */
 
-    this.cmOption.hintOptions.tables = tables;
+    this.editor.on('focus', editor => {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue:138]: Have you tried to apped this event to the textarea with `@focus`, `@change`?

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/lib/sql-theme.css (hunk 1 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/lib/sql-theme.css:8]: This class is the codemirror. I only use it to styling autocomplete

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue (hunk 7 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue]: `This.hint` activate autocomplete for SQL and Columns https://github.com/PopulateTools/gobierto/blob/902_sql_editor_autocomplete/app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue#L215

Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue]: I see. That's due to the same I tell you [here](https://github.com/PopulateTools/gobierto/pull/2843#discussion_r378160857)

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue (hunk 8 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue:138]: Yes, and it doesn't work :\

File: app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue (hunk 7 above)
Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/sets/SQLEditorCode.vue]: Even more, do not nest objects in `data()`. Every single time you do so, Vue creates observer methods for each property. In a nested object, you create observers for `hintOptions`, and then for `completeSingle`, `tables`, and so on.

That's a bad practice, who penalizes performance

**Review by entantoencuanto — DISMISSED** [review]

**Review by entantoencuanto — CHANGES_REQUESTED** [review]
If I send a query which returns empty data an exception is produced, the statistics of the query (duration and 0 rows count) are not shown and the table of the last query remains in the page
empty_response

**Comment by jorgeatgu** [comment]
> If I send a query which returns empty data an exception is produced, the statistics of the query (duration and 0 rows count) are not shown and the table of the last query remains in the page
> empty_response

This error belongs to this [branch](https://github.com/PopulateTools/gobierto/tree/908_experimental_branch_dont_merge_perspective), and it has happened when changing the display of the table 

I have openened a new issue https://github.com/PopulateTools/issues/issues/965

**Review by entantoencuanto — APPROVED** [review]

**Pull request merged**